packages/fuels/tests/bindings.rs

```diff
@@ -801,3 +806,139 @@ async fn compile_bindings_enum_input() {
     let expected = "0000000021b2784f0000000000000000000000000000002a";
     assert_eq!(encoded, expected);
 }
+
+// This is a caveat with the abigen! shared-types implementation.
+// If run inside a function, like we're doing here in this test, then the
+// normal types have issues referencing the shared types.
+//
+// A shared type, from inside a contract/script/predicate bindings mod,
+// would normally be referenced like this:
+// 'super::shared_types::SomeSharedType'. 'super' would get you out of the
+// mod of current mod and then you step down into the 'shared_types' mod.
+//
+// For some reason, 'super' inside a function gets you out of your mod, but
+// if that gets you to the create root, you can never get back in.
+//
+// A workaround is to wrap everything in another mod so that 'super' lands
+// you inside of that one and not 'outside' of the function.
+//
+// Not a big issue since almost all usages call abigen! outside of any
+// function.
```

Comment:
Although this is true, I'm 100% certain a ton of people out there will be calling `abigen!` from within functions. Just to confirm, this is only a problem if you need to navigate through the auto-generated mods to access shared types. Is that right? 

In any case, this should be super, _super_ explicit in the docs! In the same way that we encourage a single `abigen!` call, we should encourage `abigen!` outside of functions and let the user know these caveats.

Reply:
Also, this whole `shared_types()` test below is an excellent way to demonstrate how to handle type conflicts under these caveats; should we add this to a doc example? Maybe a whole sub-section about shared types?

Reply:
Oh man, I should have deleted this comment. This is no longer an issue since we decided on the meet to wrap everything in a parent mod `abigen_bindings`. 

I wanted to add this as an example but got cold feet. The type names seem to be overly verbose for documentation purposes. I named them that way to make tests more readable. 

But I guess if that makes the tests readable it should make the documentation more readable as well.

I can add it if you don't find it over the top. It handles both shared types, unique types, and types that won't get a `pub use` since they are not uniquely typed.

With some explanation about what is happening, it might just work. Lemme know.

Reply:
Will make a separate issue, as agreed on the sync.